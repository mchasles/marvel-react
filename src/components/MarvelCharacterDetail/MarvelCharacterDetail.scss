.marvel-character-detail {
  flex-grow: 1;
  padding: 20px;
  position: relative;

  h1 {
    color: #61dafb;
    font-size: 42px;
    font-weight: 100;
    margin-top: 0;
    text-align: center;
  }

  img.thumbnail {
    animation: thumbnail 0.3s 1;
    border-radius: 50%;
    box-shadow: 0 0 6px black;
    display: block;
    height: 400px;
    margin: 0 auto;
    object-fit: cover;
    width: 400px;
  }

  @keyframes thumbnail {
    0% {
      transform: scale(0);
    }
    100% {
      transform: scale(1);
    }
  }

  .img-404 {
    background-color: #61dafb;
    border-radius: 50%;
    box-shadow: 0 0 6px black;
    color: white;
    font-size: 42px;
    height: 400px;
    line-height: 400px;
    margin: 0 auto;
    text-align: center;
    width: 400px;
  }
  
}