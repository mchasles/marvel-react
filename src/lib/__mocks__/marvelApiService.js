const IMG_NOT_AVAILABLE_SRC = 'http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available/detail.jpg';

export const characters = [
  {
    'id': '1',
    'name': '3-D Man',
    'thumbnail': {
      'path': 'http://mock/1'
    }
  },
  {
    'id': '2',
    'name': 'A-Bomb (HAS)',
    'thumbnail': {
      'path': ''
    }
  }
];

export const loadCharacters = () => {
  return new Promise((resolve, reject) => {
    resolve(characters);
  });
}

export const loadCharacter = (id) => {
  return new Promise((resolve, reject) => {
    let character = characters.find(ch => ch.id === id);
    character ? resolve(character) : reject({
      error: `Character with ${id} not found.`,
    })
  });
}

export const getCharacterPortraitSrc = (path) => {
  return new Promise((resolve, reject) => {
    if (path.length) {
      resolve(path + '/portrait.jpg');
    } else {
      reject(IMG_NOT_AVAILABLE_SRC);
    }
  });
}