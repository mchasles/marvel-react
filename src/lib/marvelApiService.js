import axios from 'axios';
// API endpoint
const BASE_URL = 'https://gateway.marvel.com/v1/public/characters';
// API key
const API_KEY = 'cefd21ddeeea8bc5232fa3fe67c57e0f';
// URL of the default image when character portrait is not available
export const IMG_NOT_AVAILABLE_SRC = 'http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available/detail.jpg';
// URL suffix to add to resourceURI to get portrait
export const URL_PORTRAIT_SUFFIX  = '/detail.jpg';

// Fetch the 16 first characters from Marvel API (pagination will come later...)
export const loadCharacters = () => {
  return axios.get(BASE_URL, {
    params: {
      apikey: API_KEY,
      limit: 16,
      offset: 0
    }
  }).then(res => res.data.data.results)
};

// Fetch character info from Marvel API with its id
export const loadCharacter = (id) => {
  return axios.get(`${BASE_URL}/${id}`,{
    params: {
      apikey: API_KEY
    }
  }).then(res => res.data.data.results[0])
};

// Request character portrait image to check if available, if an error happens return default image src
export const getCharacterPortraitSrc = (path) => {
    let portraitSrc = `${path}${URL_PORTRAIT_SUFFIX}`;
    return axios.get(portraitSrc)
      .then(
        // Successfully get character portrait (but image might be the default one when not available)
        () => portraitSrc,
        // Error getting character portrait (error 404)
        () => IMG_NOT_AVAILABLE_SRC);
};