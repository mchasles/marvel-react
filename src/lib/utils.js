// Preload img in browser cache
export const preloadImg = (src) => {
  return new Promise((resolve, reject) => {
    const img = document.createElement('img');
    img.src = src;
    img.onload = () => resolve('done');
    // If image does not load after 6 seconds, reject promise
    setTimeout(() => {
      reject('error')
    }, 6000);
  });
}