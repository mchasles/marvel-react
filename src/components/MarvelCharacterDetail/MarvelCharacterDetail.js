import React, { Component } from 'react';
import {IMG_NOT_AVAILABLE_SRC, loadCharacter, getCharacterPortraitSrc} from '../../lib/marvelApiService';
import {preloadImg} from '../../lib/utils';
import {ReactLogoLoader} from '../ReactLogoLoader/';
import './MarvelCharacterDetail.scss';

export class MarvelCharacterDetail extends Component {
  
  state = {
    character: null
  }

  componentWillReceiveProps(newProps) {
    this.setState({character: null}); // Set character state to null to display loader
    return loadCharacter(newProps.match.params.id)
      .then(this.setCharacterPortraitSrc.bind(this));
  }

  setCharacterPortraitSrc(character){
    return getCharacterPortraitSrc(character.thumbnail.path)
      .then(
        // Successfully get character portrait (but image might be the default one when not available)
        (portraitSrc) => {
          // Preload image in browser's cache before updating state
          character.portraitSrc = portraitSrc;
          
          return preloadImg(character.portraitSrc)
            .then(
              () => {
                this.setState({character})
              },
              () => {
                // If image failed to load, set default image
                character.portraitSrc = IMG_NOT_AVAILABLE_SRC;
                this.setState({character});
              },
            );
        },
        // Error getting character portrait (error 404)
        (portraitSrc) => {
          character.portraitSrc = IMG_NOT_AVAILABLE_SRC;
          this.setState({character});
        });
  }

  render() {
    return <div className="marvel-character-detail">{ this.renderContent() }</div>;
  }

  renderContent() {
    const character = this.state.character;
    if (!character) {
      return <ReactLogoLoader />;
    }
    return (
      <div>
        <h1>{character.name}</h1> 
        {this.renderThumbnail()}
      </div>
    );
  }

  renderThumbnail() {
    const character = this.state.character;
    if (character.portraitSrc === IMG_NOT_AVAILABLE_SRC) {
      // If image is not available, replace default image by custom message
      return (<div className="img-404">Image Not Found</div>);
    }
    return <img className="thumbnail" src={character.portraitSrc} alt={character.name}/>;
  }

}