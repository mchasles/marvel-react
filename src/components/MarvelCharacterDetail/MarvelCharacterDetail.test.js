import React from 'react';
import ReactDOM from 'react-dom';
import {MarvelCharacterDetail} from './MarvelCharacterDetail';

import renderer from 'react-test-renderer';
jest.mock('../../lib/marvelApiService');
import {characters} from '../../lib/marvelApiService';

jest.mock('../../lib/utils');

const props = {
  character: characters[0],
  match: {
    params: {id: '1'}
  }
};

const propsError = {
  character: characters[1],
  match: {
    params: {id: '2'}
  }
};

it('renders MarvelCharacterDetail without crashing and display loader', () => {
  const loaderRendering = renderer.create(React.createElement(MarvelCharacterDetail, props));
  expect(loaderRendering.toJSON()).toMatchSnapshot();
});

it('renders MarvelCharacterDetail correctly when there is one character selected and fetching done', (done) => {
  const characterWithPortrait = renderer.create(React.createElement(MarvelCharacterDetail, props));
  // Trigger componentWillReceiveProps as if route changed successfully and props changed
  characterWithPortrait.getInstance().componentWillReceiveProps(props)
    .then(() => {
      expect(characterWithPortrait.toJSON()).toMatchSnapshot();
      done();
    });
});

it('renders MarvelCharacterDetail with error img if character portrait failed to load', (done) => {
  const characterWithNoPortrait = renderer.create(React.createElement(MarvelCharacterDetail, propsError));
  // Trigger componentWillReceiveProps as if route changed successfully and props changed
  characterWithNoPortrait.getInstance().componentWillReceiveProps(propsError)
    .then(() => {
      expect(characterWithNoPortrait.toJSON()).toMatchSnapshot();
      done();
    });
});