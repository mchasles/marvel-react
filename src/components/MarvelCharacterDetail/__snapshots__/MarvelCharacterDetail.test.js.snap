exports[`test renders MarvelCharacterDetail correctly when there is one character selected and fetching done 1`] = `
<div
  className="marvel-character-detail">
  <div>
    <h1>
      3-D Man
    </h1>
    <img
      alt="3-D Man"
      className="thumbnail"
      src="http://mock/1/portrait.jpg" />
  </div>
</div>
`;

exports[`test renders MarvelCharacterDetail with error img if character portrait failed to load 1`] = `
<div
  className="marvel-character-detail">
  <div>
    <h1>
      A-Bomb (HAS)
    </h1>
    <div
      className="img-404">
      Image Not Found
    </div>
  </div>
</div>
`;

exports[`test renders MarvelCharacterDetail without crashing and display loader 1`] = `
<div
  className="marvel-character-detail">
  <img
    alt="React logo spinning"
    className="react-logo-loader"
    src="logo.svg" />
</div>
`;
